vm/src/protocol/buffer.rs

```diff
@@ -1,119 +1,142 @@
-    /// might operate on, among other footguns.
-    fn obj_bytes_mut(&self) -> BorrowedValueMut<[u8]>;
-    fn release(&self);
-    // not included in PyBuffer protocol itself
-    fn retain(&self);
+use itertools::Itertools;
+
+use crate::{
+    common::{
+        borrow::{BorrowedValue, BorrowedValueMut},
+        lock::{MapImmutable, PyMutex, PyMutexGuard},
+    },
+    sliceable::wrap_index,
+    types::{Constructor, Unconstructible},
+    PyObject, PyObjectPayload, PyObjectRef, PyObjectView, PyObjectWrap, PyRef, PyResult,
+    TryFromBorrowedObject, TypeProtocol, VirtualMachine,
+};
+use std::{borrow::Cow, fmt::Debug, ops::Range};
+
+#[allow(clippy::type_complexity)]
+pub struct BufferMethods {
+    pub obj_bytes: fn(&PyBuffer) -> BorrowedValue<[u8]>,
+    pub obj_bytes_mut: fn(&PyBuffer) -> BorrowedValueMut<[u8]>,
+    pub release: fn(&PyBuffer),
+    pub retain: fn(&PyBuffer),
 }
 
-#[derive(Debug)]
+impl Debug for BufferMethods {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BufferMethods")
+            .field("obj_bytes", &(self.obj_bytes as usize))
+            .field("obj_bytes_mut", &(self.obj_bytes_mut as usize))
+            .field("release", &(self.release as usize))
+            .field("retain", &(self.retain as usize))
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct PyBuffer {
     pub obj: PyObjectRef,
-    pub options: BufferOptions,
-    pub(crate) internal: PyRc<dyn BufferInternal>,
+    pub desc: BufferDescriptor,
+    methods: &'static BufferMethods,
 }
 
 impl PyBuffer {
-    pub fn new(
-        obj: PyObjectRef,
-        buffer: impl BufferInternal + 'static,
-        options: BufferOptions,
-    ) -> Self {
-        buffer.retain();
-        Self {
+    pub fn new(obj: PyObjectRef, desc: BufferDescriptor, methods: &'static BufferMethods) -> Self {
+        let zelf = Self {
             obj,
-            options,
-            internal: PyRc::new(buffer),
-        }
+            desc: desc.validate(),
+            methods,
+        };
+        zelf.retain();
+        zelf
     }
+
     pub fn as_contiguous(&self) -> Option<BorrowedValue<[u8]>> {
-        if !self.options.contiguous {
-            return None;
-        }
-        Some(self.internal.obj_bytes())
+        self.desc
+            .is_contiguous()
+            .then(|| unsafe { self.contiguous() })
     }
 
     pub fn as_contiguous_mut(&self) -> Option<BorrowedValueMut<[u8]>> {
-        if !self.options.contiguous {
-            return None;
-        }
-        Some(self.internal.obj_bytes_mut())
+        (!self.desc.readonly && self.desc.is_contiguous()).then(|| unsafe { self.contiguous_mut() })
     }
 
-    pub fn to_contiguous(&self) -> Vec<u8> {
-        self.internal.obj_bytes().to_vec()
+    /// # Safety
+    /// assume the buffer is contiguous
+    pub unsafe fn contiguous(&self) -> BorrowedValue<[u8]> {
```

Comment:
if we are going to name this under rust convention,
```suggestion
    pub unsafe fn contiguous_unchecked(&self) -> BorrowedValue<[u8]> {
```